src/mds/FSMap.cc

```diff
@@ -0,0 +1,762 @@
+        f->dump_unsigned("rank", info.rank);
+        f->dump_string("name", info.name);
+        f->dump_string("status", s);
+        f->close_section();
+      } else {
+        by_rank[mds_role_t(fscid, info.rank)] = info.name + "=" + s;
+      }
+    } else {
+      by_state[s]++;
+    }
+  }
+
+  if (f) {
+    f->close_section();
+  } else {
+    if (!by_rank.empty()) {
+      if (filesystems.size() > 1) {
+        // Disambiguate filesystems
+        std::map<std::string, std::string> pretty;
+        for (auto i : by_rank) {
+          const auto &fs_name = filesystems.at(i.first.fscid)->mds_map.fs_name;
+          std::ostringstream o;
+          o << "[" << fs_name << ":" << i.first.rank << "]";
+          pretty[o.str()] = i.second;
+        }
+        *out << " " << pretty;
+      } else {
+        *out << " " << by_rank;
+      }
+    }
+  }
+
+  for (map<string,int>::reverse_iterator p = by_state.rbegin(); p != by_state.rend(); ++p) {
+    if (f) {
+      f->dump_unsigned(p->first.c_str(), p->second);
+    } else {
+      *out << ", " << p->second << " " << p->first;
+    }
+  }
+
+  size_t failed = 0;
+  size_t damaged = 0;
+  for (auto i : filesystems) {
+    auto fs = i.second;
+    failed += fs->mds_map.failed.size();
+    damaged += fs->mds_map.damaged.size();
+  }
+
+  if (failed > 0) {
+    if (f) {
+      f->dump_unsigned("failed", failed);
+    } else {
+      *out << ", " << failed << " failed";
+    }
+  }
+
+  if (damaged > 0) {
+    if (f) {
+      f->dump_unsigned("damaged", damaged);
+    } else {
+      *out << ", " << damaged << " damaged";
+    }
+  }
+  //if (stopped.size())
+  //out << ", " << stopped.size() << " stopped";
+}
+
+void FSMap::get_health(list<pair<health_status_t,string> >& summary,
+			list<pair<health_status_t,string> > *detail) const
+{
+  for (auto i : filesystems) {
+    auto fs = i.second;
+
+    // TODO: move get_health up into here so that we can qualify
+    // all the messages with what filesystem they're talking about
+    fs->mds_map.get_health(summary, detail);
+  }
+}
+
+void FSMap::encode(bufferlist& bl, uint64_t features) const
+{
+  ENCODE_START(10, 10, bl);
+  ::encode(epoch, bl);
+  ::encode(next_filesystem_id, bl);
+  ::encode(legacy_client_fscid, bl);
+  ::encode(compat, bl);
+  ::encode(enable_multiple, bl);
+  std::vector<Filesystem> fs_list;
+  for (auto i : filesystems) {
+    fs_list.push_back(*(i.second));
+  }
+  ::encode(fs_list, bl);
+  ::encode(mds_roles, bl);
+  ::encode(standby_daemons, bl, features);
+  ::encode(standby_epochs, bl);
+  ENCODE_FINISH(bl);
+}
+
+void FSMap::decode(bufferlist::iterator& p)
+{
+  // Because the mon used to store an MDSMap where we now
+  // store an FSMap, FSMap knows how to decode the legacy
+  // MDSMap format (it never needs to encode it though).
+  Filesystem legacy_fs;
+  MDSMap &legacy_mds_map = legacy_fs.mds_map;
+  bool enabled = false;
+  
+  DECODE_START_LEGACY_COMPAT_LEN_16(10, 4, 4, p);
+  if (struct_v < 10) {
```

Comment:
Can you do a follow-on PR that moves the legacy decoding into its own function? That makes it a lot easier to match up with the encode function and keeps us from accidentally editing the legacy decoding bits.